lsp/nls/src/server.rs

```diff
@@ -0,0 +1,124 @@
+        cache.load_stdlib().unwrap();
+        Server { connection, cache }
+    }
+
+    pub(crate) fn reply(&mut self, response: Response) {
+        trace!("Sending response: {:#?}", response);
+        self.connection
+            .sender
+            .send(Message::Response(response))
+            .unwrap();
+    }
+    pub(crate) fn notify(&mut self, notification: Notification) {
+        trace!("Sending notification: {:#?}", notification);
+        self.connection
+            .sender
+            .send(Message::Notification(notification))
+            .unwrap();
+    }
+
+    fn err<E>(&mut self, id: RequestId, err: E)
+    where
+        E: std::fmt::Display,
+    {
+        warn!("{}", err);
+        self.reply(Response::new_err(
+            id,
+            ErrorCode::UnknownErrorCode as i32,
+            err.to_string(),
+        ));
+    }
+
+    pub fn run(&mut self) -> Result<()> {
+        trace!("Running...");
+        while let Ok(msg) = self.connection.receiver.recv() {
+            trace!("Message: {:#?}", msg);
+            match msg {
+                Message::Request(req) => {
+                    let id = req.id.clone();
+                    match self.connection.handle_shutdown(&req) {
+                        Ok(true) => break,
+                        Ok(false) => self.handle_request(req),
+                        Err(err) => {
+                            // This only fails if a shutdown was
+                            // requested in the first place, so it
+                            // should definitely break out of the
+                            // loop.
+                            self.err(id, err);
+                            break;
+                        }
+                    }
+                }
+                Message::Notification(notification) => {
+                    let _ = self.handle_notification(notification);
+                }
+                Message::Response(_) => (),
+            }
+        }
+
+        Ok(())
+    }
+
+    fn handle_notification(&mut self, notification: Notification) {
+        match notification.method.as_str() {
+            DidOpenTextDocument::METHOD => {
+                trace!("handle open notification");
+                crate::files::handle_open(
+                    self,
+                    serde_json::from_value::<DidOpenTextDocumentParams>(notification.params)
+                        .unwrap(),
+                )
+                .unwrap();
+            }
+            DidChangeTextDocument::METHOD => {
+                trace!("handle save notification");
+                crate::files::handle_save(
+                    self,
+                    serde_json::from_value::<DidChangeTextDocumentParams>(notification.params)
+                        .unwrap(),
+                )
+                .unwrap();
+            }
+            _ => {}
+        }
+    }
+
+    fn handle_request(&self, req: lsp_server::Request) {
+        todo!()
```

Comment:
I didn't even know this existed. I would have used `unimplemented!()`, which is fine I guess, but still it's good to know :)